feat: translations via gettext

barredterra opened the pull request:
- Move navbar and help items to hooks in order to be able to extract them for translations
- Translatable string extractors:
    - Python files (*.py)
    - JavaScript files (*.js)
    - Jinja2 templates (*.html)
    - DocType JSON
    - Report JSON
    - Workspace JSON
    - Navbar and help items (from hooks.py)
    - Module Onboarding
    - Onboarding Step
- Folder structure
    - `[app]/locale/main.pot` holds all translatable strings of an app, along with context and comments
    - `[app]/locale/[locale].po` a copy of `main.pot`, with a subset translated to this language
    - `sites/assets/locale/[locale]/LC_MESSAGES/[app].mo` holds all translations of an app into a specific language, compiled into a binary format for optimized retrieval
- Utils for creating `.pot` -> `.po` -> `.mo` files
- Bench commands to access above utils
    - `generate-pot-file` [generate a template file with all of an app's translatable strings](https://frappeframework.com/docs/user/en/bench-generate-pot-file)
    - `migrate-csv-to-po` [migrate translations from CSV files (old) to PO files (new)](https://frappeframework.com/docs/user/en/bench-migrate-csv-to-po)
    - `update-po-files` [sync PO files with POT file](https://frappeframework.com/docs/user/en/bench-update-po-files)
    - `create-po-file` create a new PO file for a specific locale
    - `compile-po-to-mo` [compile PO files to MO files](https://frappeframework.com/docs/user/en/bench-compile-po-to-mo)

This PR is similar to https://github.com/frappe/frappe/pull/19655 (some code copied, some added), but tries to do these things differently:

- Non-breaking. For now we just add stuff, to ensure a smooth transition.
- Smaller files. For example, each extractor gets its own file.
- No bulk-changes to translations, in order to keep the PR reviewable.

Todo:

- [x] put `.mo` files into `sites/assets/locale/[lang]/LC_MESSAGES/[app].mo`
- [x] enable translation functions to optionally use the new files 

    Load translations from both `.csv` and `.mo` files, whatever is available.

- [x] ~~migration of sub-languages, for example es_MX. Maybe copy `es.po`, then update from `es-mx.csv`.~~

    No need. We can just copy sub-languages as is. Only during the final translation, we will coalesce between sub-languages and general translation.

- [x] add extractor for **Module Onboarding** and **Onboarding Step**
- [x] apps can provide their own extractors, for files not covered by frappe's standard extractors

    Each file gets processed only by the first extraction method that matches the file pattern. Apps can place a custom `babel_extractors.csv` in their root directory. These take precedence, if any. The default extractors from frappe's `babel_extractors.csv` get appended (lower relevance).

- [x] add tests

- [x] compile translations during `bench update` (takes ~10 seconds each, for frappe and erpnext) or try to get away with not compiling anything (just load the `.po` files)
- [x] add docs (after merge)
- [x] ~~reduce code duplication (see if some existing extractors can be merged with new extractors)~~ (can be done later)

Backport to `version-15`: https://github.com/frappe/frappe/pull/25982

barredterra committed:
refactor: move navbar and help items to hooks

barredterra committed:
feat: python extractor

barredterra committed:
feat: javascript extractor

barredterra committed:
feat: doctype extractor

barredterra committed:
feat: jinja2 extractor

barredterra committed:
feat: navbar extractor

barredterra committed:
feat: report extractor

barredterra committed:
feat: workspace extractor

barredterra committed:
docs: extractor README

barredterra committed:
feat: gettext utils

barredterra committed:
feat: gettext bench commands

barredterra committed:
fix: navbar extractor

barredterra committed:
refactor: folder structure

- put .po files into [app]/locale/[lang].po
- put .mo files into sites/assets/locale/[lang]/LC_MESSAGES/[app].mo

The latter structure is mandated by the gettext library.

barredterra committed:
fix: consistent naming `migrate-csv-to-po`

barredterra committed:
Merge remote-tracking branch 'upstream/develop' into po-translation

barredterra committed:
chore: remove unused methods

These are much slower than the other approach (results are not cached).
Also, they are lacking a fallback mechanism for missing context.

barredterra committed:
chore: use existing get_dict_from_hooks

barredterra committed:
fix: rename cache keys to avoid conflict

barredterra committed:
fix: get_translations_from_apps

- Only consider installed apps
- Handle context correctly

barredterra committed:
feat: optionally use gettext for translation

barredterra committed:
feat: add extractor for Onboarding Step

barredterra committed:
feat: add extractor for Module Onboarding

barredterra committed:
Merge branch 'develop' into po-translation

barredterra committed:
chore: remove unused method

barredterra committed:
fix: prefer more specific extractors

barredterra committed:
fix: integrate upstream changes

barredterra committed:
refactor: reuse existing `get_user_translations`

barredterra committed:
fix: integrate upstream changes

barredterra committed:
fix: conf must be available

ankush committed:
Merge branch 'develop' into po-translation

barredterra committed:
Merge remote-tracking branch 'upstream/develop' into po-translation

barredterra committed:
feat: support CSV and gettext at the same time

barredterra committed:
fix: clear translation cache

- Cache was not getting cleared, since it's user-specific
- Remove unused implementation from gettext

barredterra committed:
fix: don't presume english as source language

Note: If we DO want this, it should happen at a higher level.

barredterra committed:
feat: apps can provide custom message extractors

- custom extractors take precedence
- default extractors are used as a fallback

ankush commented on babel_extractors.csv:
```diff
@@ -0,0 +1,9 @@
+hooks.py,frappe.gettext.extractors.navbar.extract
```
Why CSV format for configuration? 

hooks.py or pyproject.toml is better for this IMO.

ankush replied:
Is this because you don't want to rely on DB connection?

barredterra replied:
Yes, reading `hooks.py` currently depends on a db connection and I wanted to avoid this. Reading `toml` files seemed to require an additional library. So I thought let's keep it simple and go with CSV.

ankush commented on frappe/commands/utils.py:
```diff
@@ -102,6 +103,7 @@ def clear_cache(context):
 			frappe.clear_cache()
 			clear_notifications()
 			clear_website_cache()
+			clear_translations()
```
probably not required. `frappe.clear_cache()` clears EVERYTHING now. 

https://github.com/frappe/frappe/blob/7d4bdbebb9c80742e098b97bb1008fdb8c340fcc/frappe/__init__.py#L963-L964 


I got tired of weird cache_manager tracking and just added this nuclear option when no args are passed to frappe.clear_cache.

barredterra replied:
Huh, I was under the impression that `bench clear-cache` didn't clear the translations cache. 🤔

ankush commented on frappe/gettext/extractors/javascript.py:
```diff
@@ -0,0 +1,26 @@
+from babel.messages.extract import extract_javascript
+
+
+def extract(fileobj, keywords, comment_tags, options):
```
@barredterra AFAIK extract_javascript from babel doesn't work for us. Hence I copied it and modified it to work with Frappe codebase. 

You can find it here: https://github.com/frappe/frappe/blob/7d4bdbebb9c80742e098b97bb1008fdb8c340fcc/frappe/translate.py#L678-L701

It will have to again be modified (or just wrapped) for it to work with babel interface.

ankush replied:
If all of this is somehow handled then let me know :eyes:

ankush committed:
chore: remove duplicate cache clearing

frappe.clear_cache removes everything

barredterra committed:
test: add cases for js parser

barredterra committed:
fix: move `extract_javascript` to extractors folder and use it

instead of babels js extractor

barredterra committed:
fix: handle missing locale directory

barredterra committed:
fix: avoid writing empty .mo files

barredterra committed:
fix: variable name conflict

barredterra committed:
refactor: variable names

barredterra committed:
test: gettext utils

ankush committed:
Merge branch 'develop' into po-translation

ankush committed:
feat: Build MO files with `bench build`

Translation assets are also... assets, so grouping them with `bench build`.

We can do some special function but that will likely require updating
bench for everyone. Not so ideal.

Also renamed `compile` function as it is a builtin in python.

ankush committed:
perf: Skip rebuilding MO files where PO file is not updated

ankush committed:
perf: parallelize MO compilation

This operation is CPU heavy and has little to no Frappe code
involvement. So we can use a processpool of 4 to reduce time taken for
compilation by 4x.

ankush commented:
Made few changes:
- `bench build` now builds translation assets
- We skip PO files where MO file is newer than PO file during compilation
- Compilation happens in processpool of 4, almost cuts down time by 1/4th. I tested out different processpool sizes. 4 seems to be best, more than 4 have steep diminishing returns. 
- Migrated CSV to PO files.

```
λ time bench compile-po-to-mo --app frappe --force > /dev/null
bench compile-po-to-mo --app frappe --force > /dev/null  11.25s user 0.81s system 100% cpu 12.053 total
λ gco -
Previous HEAD position was 9482126b93 perf: Skip rebuilding MO files where PO file is not updated
Switched to branch 'po-translation'
λ time bench compile-po-to-mo --app frappe --force > /dev/null
bench compile-po-to-mo --app frappe --force > /dev/null  12.07s user 0.83s system 279% cpu 4.608 total
```

ankush committed:
chore: migrate translations to PO files

ankush committed:
fix: language variants support

ankush commented:
@barredterra Looks good for first-cut merge from my side. :+1: 

We can do subsequent changes if any post-merge (?)

ankush commented:
note: Port with all fixes linked to https://github.com/frappe/frappe/issues/24250

ankush commented:
@mergifyio backport version-15-hotfix

ankush commented:
Bad bot
